docs/src/pages/components/data-grid/state/RestoreStateApiRef.tsx

```diff
@@ -0,0 +1,334 @@
+      >
+        <TextField
+          value={label}
+          onChange={onLabelChange}
+          size="small"
+          label="Custom view label"
+          variant="standard"
+        />
+      </ListItem>
+    );
+  }
+
+  return (
+    <ListItem>
+      <Button
+        size="small"
+        startIcon={<AddIcon />}
+        onClick={() => setIsAddingView(true)}
+      >
+        Add a custom view
+      </Button>
+    </ListItem>
+  );
+};
+
+const CustomToolbar = () => {
+  const apiRef = useGridApiContext();
+  const [state, dispatch] = React.useReducer(demoReducer, DEMO_INITIAL_STATE);
+
+  const createNewView = () => {
+    dispatch({
+      type: 'createView',
+      value: apiRef.current.exportState(),
+    });
+  };
+
+  const handleNewViewLabelChange = (
+    e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
+  ) => {
+    dispatch({ type: 'setNewViewLabel', label: e.target.value });
+  };
+
+  const handleDeleteView = React.useCallback((viewId: string) => {
+    dispatch({ type: 'deleteView', id: viewId });
+  }, []);
+
+  const handleSetActiveView = (viewId: string) => {
+    apiRef.current.restoreState(state.views[viewId].value);
+    dispatch({ type: 'setActiveView', id: viewId });
+  };
+
+  const handlePopperAnchorClick = (event: React.MouseEvent) => {
+    dispatch({ type: 'togglePopper', element: event.currentTarget as HTMLElement });
+    event.stopPropagation();
+  };
+
+  const handleClosePopper = () => {
+    dispatch({ type: 'closePopper' });
+  };
+
+  const isNewViewLabelValid = React.useMemo(() => {
+    if (state.newViewLabel.length === 0) {
+      return false;
+    }
+
+    return Object.values(state.views).every(
+      (view) => view.label !== state.newViewLabel,
+    );
+  }, [state.views, state.newViewLabel]);
+
+  const canBeMenuOpened = state.isMenuOpened && Boolean(state.menuAnchorEl);
+  const popperId = canBeMenuOpened ? 'transition-popper' : undefined;
+
+  return (
+    <React.Fragment>
+      <Button
+        aria-describedby={popperId}
+        type="button"
+        onClick={handlePopperAnchorClick}
+      >
+        Custom views
+      </Button>
+      <ClickAwayListener onClickAway={handleClosePopper}>
+        <Popper
+          id={popperId}
+          open={state.isMenuOpened}
+          anchorEl={state.menuAnchorEl}
+          transition
+          placement="bottom-start"
+        >
+          {({ TransitionProps }) => (
+            <Fade {...TransitionProps} timeout={350}>
+              <Paper>
+                <List>
+                  {Object.entries(state.views).map(([viewId, view]) => (
+                    <ViewListItem
+                      key={viewId}
+                      view={view}
+                      viewId={viewId}
+                      selected={viewId === state.activeViewId}
+                      onDelete={handleDeleteView}
+                      onSelect={handleSetActiveView}
+                    />
+                  ))}
+                  <Divider />
```

Comment:
Something is odd with this demo. Having the text input inside the list doesn't look very good. What about putting the input outside the menu, like in a toolbar?

![image](https://user-images.githubusercontent.com/42154031/151804977-cb32b525-c904-4d95-9127-f75627bb53aa.png)

Because of the way this variant works it appears misaligned when not fucused.

![image](https://user-images.githubusercontent.com/42154031/151805239-10b1546c-5421-49c1-89b2-ddcb20fa4d8a.png)

Reply:
I did not finalize this demo, as I wanted feedbacks about the global idea before spending to much times on the details.
But I agree that it is not 100% clean right now :+1: